File: engine/src/graphics/tim2.cpp
#include "../include/graphics/tim2.h"

#include <cstring>

#include "EngineDebug.h"

namespace
{
// Alignment-safe little-endian reads (the blob comes straight off disc and its
// multi-byte fields are not guaranteed to sit on aligned addresses).
uint16_t ReadU16(const uint8_t* p)
{
    uint16_t v;
    std::memcpy(&v, p, sizeof(v));
    return v;
}

uint32_t ReadU32(const uint8_t* p)
{
    uint32_t v;
    std::memcpy(&v, p, sizeof(v));
    return v;
}

// File header layout (16 bytes):
//   [0..3]  magic "TIM2"   [4] formatVersion   [5] formatId
//   [6..7]  pictureCount   [8..15] padding
constexpr uint32_t TIM2_MAGIC = 0x324D4954u; // 'T','I','M','2' little-endian
constexpr size_t TIM2_FILE_HEADER_SIZE = 16;

// Picture header layout (48 bytes / 0x30) — see tim2.h. We only read what we
// need: sizes, dimensions and the image color type.
constexpr size_t TIM2_PIC_HEADER_SIZE = 0x30;

// TIM2 imageType values we accept.
constexpr uint8_t TIM2_IMGTYPE_RGBA16 = 0x01; // A1B5G5R5
constexpr uint8_t TIM2_IMGTYPE_RGBA32 = 0x03; // A8B8G8R8
constexpr uint8_t TIM2_IMGTYPE_IDTEX8 = 0x05; // 8-bit indexed + CLUT

inline size_t Align16(size_t n) { return (n + 15u) & ~static_cast<size_t>(15u); }
} // namespace

bool Tim2_Parse(const void* data, size_t size, Tim2Image* out)
{
    if (!data || !out || size < TIM2_FILE_HEADER_SIZE)
    {
        Engine_LogError("TIM2: null/short blob (%zu bytes).", size);
        return false;
    }

    const uint8_t* base = static_cast<const uint8_t*>(data);

    if (ReadU32(base) != TIM2_MAGIC)
    {
        Engine_LogError("TIM2: bad magic.");
        return false;
    }

    const uint8_t formatId = base[5];
    const uint16_t pictureCount = ReadU16(base + 6);
    if (pictureCount == 0)
    {
        Engine_LogError("TIM2: zero pictures.");
        return false;
    }

    // pack_assets.py always writes formatId 0 (16-byte alignment → picture header
    // immediately after the 16-byte file header). formatId 1 aligns to 128.
    const size_t picBase = (formatId == 0) ? TIM2_FILE_HEADER_SIZE : 128;
    if (size < picBase + TIM2_PIC_HEADER_SIZE)
    {
        Engine_LogError("TIM2: truncated picture header.");
        return false;
    }

    const uint8_t* pic = base + picBase;
    const uint32_t clutSize = ReadU32(pic + 4);
    const uint32_t imageSize = ReadU32(pic + 8);
    const uint16_t headerSize = ReadU16(pic + 12);
    const uint8_t mipmapCount = pic[17];
    const uint8_t imageType = pic[19];
    const uint16_t imageWidth = ReadU16(pic + 20);
    const uint16_t imageHeight = ReadU16(pic + 22);

    PixelFormat fmt;
    uint32_t bpp; // bytes per texel for the image levels
    switch (imageType)
    {
    case TIM2_IMGTYPE_RGBA32:
        fmt = PixelFormat::RGBA32;
        bpp = 4;
        break;
    case TIM2_IMGTYPE_RGBA16:
        fmt = PixelFormat::RGBA16;
        bpp = 2;
        break;
    case TIM2_IMGTYPE_IDTEX8:
        fmt = PixelFormat::PAL8;
        bpp = 1;
        break;
    default:
        Engine_LogError("TIM2: unsupported imageType 0x%02X.", imageType);
        return false;
    }

    const uint8_t mipCount = (mipmapCount == 0) ? 1 : mipmapCount;
    if (mipCount > TEX_MAX_MIP_LEVELS)
    {
        Engine_LogError("TIM2: too many mip levels (%u > %d).", mipCount, TEX_MAX_MIP_LEVELS);
        return false;
    }
    if (fmt == PixelFormat::PAL8 && clutSize < 256u * 4u)
    {
        Engine_LogError("TIM2: PAL8 image missing CLUT (clutSize=%u).", clutSize);
        return false;
    }

    // Image data starts headerSize bytes into the picture block. Levels are
    // stored contiguously largest-first, each 16-byte aligned (see pack_assets.py).
    const size_t imgOffset = picBase + headerSize;
    if (headerSize < TIM2_PIC_HEADER_SIZE || imgOffset + imageSize > size)
    {
        Engine_LogError("TIM2: image payload out of bounds (hdr=%u, size=%u).", headerSize, imageSize);
        return false;
    }

    std::memset(out->levelPtr, 0, sizeof(out->levelPtr));
    size_t off = 0;
    for (uint8_t lvl = 0; lvl < mipCount; ++lvl)
    {
        const uint32_t w = (imageWidth >> lvl) ? static_cast<uint32_t>(imageWidth >> lvl) : 1u;
        const uint32_t h = (imageHeight >> lvl) ? static_cast<uint32_t>(imageHeight >> lvl) : 1u;
        if (off + static_cast<size_t>(w) * h * bpp > imageSize)
        {
            Engine_LogError("TIM2: mip level %u out of bounds.", lvl);
            return false;
        }
        out->levelPtr[lvl] = base + imgOffset + off;
        off += Align16(static_cast<size_t>(w) * h * bpp);
    }

    // CLUT (PAL8) follows the image payload.
    out->clut = nullptr;
    if (fmt == PixelFormat::PAL8)
    {
        const size_t clutOffset = imgOffset + imageSize;
        if (clutOffset + 256u * 4u > size)
        {
            Engine_LogError("TIM2: CLUT out of bounds.");
            return false;
        }
        out->clut = base + clutOffset;
    }

    out->pixels = out->levelPtr[0];
    out->mipCount = mipCount;
    out->width = imageWidth;
    out->height = imageHeight;
    out->format = fmt;
    out->imageSize = imageSize;
    return true;
}

File: tools/ps2lib/tim2.py
"""TIM2 texture encoding (PS2-native). rgba32 / pal8, optional box-filtered mips.

Byte-compatible with the historical cooker encoders (golden-tested via
BOX.PS2A). `encode_pal8_cutout` is a level-pipeline addition: it reserves CLUT
index 0 as fully transparent for billboard-impostor cutouts.
"""

import struct

TIM2_IMGTYPE_RGBA16 = 0x01
TIM2_IMGTYPE_RGBA32 = 0x03
TIM2_IMGTYPE_IDTEX8 = 0x05

# GsTex1 carries the sampling filter. Zero has always meant "let the backend
# choose", and every texture cooked before this field was used contains zero, so
# an explicit choice is marked rather than encoded as a value. Bit 63 is unused
# by the hardware register, which is what makes it available to say "the filter
# fields below are authoritative".
TEX1_EXPLICIT = 1 << 63
TEX1_MMAG_SHIFT = 8
TEX1_MMIN_SHIFT = 9
FILTER_NEAREST = 0
FILTER_LINEAR = 1


def tex1_for_filter(name):
    """Encode a filter name into a GsTex1 image, or 0 for the backend default."""
    if not name:
        return 0
    key = str(name).lower()
    if key == "nearest":
        return TEX1_EXPLICIT | (FILTER_NEAREST << TEX1_MMAG_SHIFT) | (FILTER_NEAREST << TEX1_MMIN_SHIFT)
    if key == "linear":
        return TEX1_EXPLICIT | (FILTER_LINEAR << TEX1_MMAG_SHIFT) | (FILTER_LINEAR << TEX1_MMIN_SHIFT)
    raise ValueError("unknown texture filter '%s'" % name)


def filter_from_tex1(tex1):
    """Decode a GsTex1 image; returns 'nearest', 'linear', or None for default."""
    if not (tex1 & TEX1_EXPLICIT):
        return None
    return "linear" if ((tex1 >> TEX1_MMAG_SHIFT) & 1) else "nearest"


def mip_level_count(w, h, requested):
    """Total mip levels (>=1) for `requested` extra levels, clamped so the
    smallest level stays >= 8x8."""
    levels = 1
    while levels <= requested:
        if (w >> levels) < 8 or (h >> levels) < 8:
            break
        levels += 1
    return levels


def assemble_tim2(width, height, level_payloads, image_type, clut_bytes, gs_tex1=0):
    """Assemble a single-picture TIM2 from per-level pixel payloads (largest
    first, each padded to 16 bytes) and an optional linear CLUT."""
    image = bytearray()
    for lp in level_payloads:
        image += lp
        image += b"\x00" * ((-len(image)) % 16)  # pad each level to 16 bytes
    image_size = len(image)
    clut_size = len(clut_bytes) if clut_bytes else 0
    clut_colors = (clut_size // 4) if clut_bytes else 0
    mip_count = len(level_payloads)
    clut_type = 0x03 if clut_bytes else 0  # A8B8G8R8 CLUT, CSM1, stored linear

    total_size = 0x30 + image_size + clut_size
    pic = struct.pack("<III", total_size, clut_size, image_size)  # totalSize, clutSize, imageSize
    pic += struct.pack("<HH", 0x30, clut_colors)                  # headerSize, clutColors
    pic += struct.pack("<BBBB", 0, mip_count, clut_type, image_type)  # pictFormat, mipmapCount, clutType, imageType
    pic += struct.pack("<HH", width, height)                      # imageWidth, imageHeight
    pic += struct.pack("<QQ", 0, gs_tex1)                         # GsTex0, GsTex1
    pic += struct.pack("<II", 0, 0)                               # GsRegs, GsTexClut
    assert len(pic) == 0x30, len(pic)

    # File header (16 bytes): magic, formatVersion=4, formatId=0, pictureCount=1, pad[8]
    fh = b"TIM2" + struct.pack("<BBH", 0x04, 0x00, 1) + (b"\x00" * 8)
    assert len(fh) == 16, len(fh)

    return fh + pic + bytes(image) + (clut_bytes if clut_bytes else b"")


def encode_rgba32(img, mip_levels, gs_tex1=0):
    """RGBA32 (A8B8G8R8) TIM2, box-filtered mip chain."""
    from PIL import Image
    w, h = img.size
    levels = mip_level_count(w, h, mip_levels)
    payloads = []
    for l in range(levels):
        lw, lh = max(1, w >> l), max(1, h >> l)
        lvl = img if l == 0 else img.resize((lw, lh), Image.BOX)
        payloads.append(lvl.tobytes())
    return assemble_tim2(w, h, payloads, TIM2_IMGTYPE_RGBA32, None, gs_tex1)


def encode_pal8(img, mip_levels, gs_tex1=0):
    """8-bit indexed (IDTEX8) TIM2 with a linear 256-entry A8B8G8R8 CLUT. Level 0
    is quantized; lower mips are box-filtered then remapped to the same palette
    so one CLUT serves the whole chain. Alpha is not preserved (set opaque)."""
    from PIL import Image
    w, h = img.size
    rgb = img.convert("RGB")
    pal_img = rgb.quantize(colors=256, method=Image.Quantize.FASTOCTREE)

    levels = mip_level_count(w, h, mip_levels)
    payloads = []
    for l in range(levels):
        lw, lh = max(1, w >> l), max(1, h >> l)
        idx_img = pal_img if l == 0 else rgb.resize((lw, lh), Image.BOX).quantize(palette=pal_img, dither=Image.Dither.NONE)
        payloads.append(idx_img.tobytes())

    # CLUT: palette RGB triples -> linear A8B8G8R8 (alpha 0x80 == GS 1.0), 256 entries.
    palette = pal_img.getpalette() or []
    clut = bytearray()
    for i in range(256):
        r = palette[i * 3 + 0] if i * 3 + 2 < len(palette) else 0
        g = palette[i * 3 + 1] if i * 3 + 2 < len(palette) else 0
        b = palette[i * 3 + 2] if i * 3 + 2 < len(palette) else 0
        clut += struct.pack("<BBBB", r, g, b, 0x80)
    return assemble_tim2(w, h, payloads, TIM2_IMGTYPE_IDTEX8, bytes(clut), gs_tex1)


def encode_coverage8(img, gs_tex1=0):
    """IDTEX8 TIM2 holding coverage only: the index is the source alpha and the
    CLUT is a ramp from transparent to opaque white.

    One byte per texel with antialiased edges intact, and it round-trips exactly
    through the runtime's console-alpha rescale, so the same payload is correct
    whether a backend samples it natively or expands it. This is what a glyph
    atlas wants: colour comes from the drawing call, never from the font."""
    rgba = img.convert("RGBA")
    w, h = rgba.size
    payload = rgba.getchannel("A").tobytes()

    clut = bytearray()
    for i in range(256):
        clut += struct.pack("<BBBB", 255, 255, 255, (i * 0x80) // 255)
    return assemble_tim2(w, h, [bytes(payload)], TIM2_IMGTYPE_IDTEX8, bytes(clut), gs_tex1)


def encode_pal8_cutout(img):
    """IDTEX8 TIM2 whose CLUT index 0 is fully transparent (alpha 0). Used for
    billboard-impostor atlases: fully-transparent source pixels map to index 0,
    opaque pixels are quantized into indices 1..255. No mips (impostors are drawn
    at roughly one screen size). CLUT alpha is 0x80 (GS 1.0) for opaque entries."""
    from PIL import Image
    w, h = img.size
    rgba = img.convert("RGBA")
    px = rgba.load()

    opaque_rgb = rgba.convert("RGB")
    # Quantize the RGB into 255 colours (indices become 0..254); shift to 1..255
    # so index 0 is free for transparency.
    pal_img = opaque_rgb.quantize(colors=255, method=Image.Quantize.FASTOCTREE)
    pal_idx = pal_img.load()

    indices = bytearray(w * h)
    for y in range(h):
        for x in range(w):
            if px[x, y][3] < 128:
                indices[y * w + x] = 0  # transparent
            else:
                indices[y * w + x] = pal_idx[x, y] + 1

    palette = pal_img.getpalette() or []
    clut = bytearray()
    clut += struct.pack("<BBBB", 0, 0, 0, 0)  # index 0 = transparent
    for i in range(255):
        r = palette[i * 3 + 0] if i * 3 + 2 < len(palette) else 0
        g = palette[i * 3 + 1] if i * 3 + 2 < len(palette) else 0
        b = palette[i * 3 + 2] if i * 3 + 2 < len(palette) else 0
        clut += struct.pack("<BBBB", r, g, b, 0x80)
    return assemble_tim2(w, h, [bytes(indices)], TIM2_IMGTYPE_IDTEX8, bytes(clut))


def convert_texture_to_tim2(source_path, fmt="rgba32", mip_levels=0, tex_filter=None):
    """Decode any Pillow-supported image and re-encode as TIM2. `fmt` is
    'rgba32' (default), 'pal8' or 'coverage8'; `mip_levels` is the number of
    extra mip levels; `tex_filter` is 'nearest', 'linear' or None for the
    backend default.
    Returns (tim2_bytes, ".tm2") or raises ImportError if Pillow is absent."""
    from PIL import Image
    img = Image.open(source_path).convert("RGBA")
    gs_tex1 = tex1_for_filter(tex_filter)
    if fmt == "coverage8":
        return encode_coverage8(img, gs_tex1), ".tm2"
    if fmt == "pal8":
        return encode_pal8(img, mip_levels, gs_tex1), ".tm2"
    return encode_rgba32(img, mip_levels, gs_tex1), ".tm2"


# Names for the imageType values the encoders above write. Derived from the
# constants rather than restated, so the two cannot drift.
IMAGE_TYPE_NAMES = {
    TIM2_IMGTYPE_RGBA16: "rgba16",
    TIM2_IMGTYPE_RGBA32: "rgba32",
    TIM2_IMGTYPE_IDTEX8: "pal8",
}


def describe(blob):
    """Summarise a TIM2 blob for the inspection and validation tools.

    Returns a dict, or raises ValueError. Reads only the file and picture
    headers - it never decodes pixels, so it stays cheap on a whole tree.
    """
    if len(blob) < 16 + 0x30:
        raise ValueError("shorter than a TIM2 header")
    if blob[:4] != b"TIM2":
        raise ValueError("bad TIM2 magic")

    total_size, clut_size, image_size = struct.unpack_from("<III", blob, 16)
    header_size, clut_colors = struct.unpack_from("<HH", blob, 16 + 12)
    _pict_format, mip_count, clut_type, image_type = struct.unpack_from("<BBBB", blob, 16 + 16)
    width, height = struct.unpack_from("<HH", blob, 16 + 20)
    (gs_tex1,) = struct.unpack_from("<Q", blob, 16 + 32)

    return {
        "width": width,
        "height": height,
        "format": IMAGE_TYPE_NAMES.get(image_type, "type%d" % image_type),
        "image_type": image_type,
        "mip_count": mip_count,
        "clut_colors": clut_colors,
        "clut_size": clut_size,
        "clut_type": clut_type,
        "image_size": image_size,
        "total_size": total_size,
        "header_size": header_size,
        "filter": filter_from_tex1(gs_tex1),
    }


def describe_text(blob):
    d = describe(blob)
    clut = ", clut %d colours" % d["clut_colors"] if d["clut_colors"] else ""
    filt = ", %s" % d["filter"] if d["filter"] else ""
    return "%dx%d %s, %d mip(s), %d image bytes%s%s" % (
        d["width"], d["height"], d["format"], d["mip_count"], d["image_size"], clut, filt)
